proposals/semi-auto-properties.md

```diff
@@ -10,6 +10,257 @@ Two common scenarios are that you want to apply a constraint on the setter, ensu
+        P4 = 4; // Calls P4's setter
+    }
+
+    public int P1 => field;
+    public int P2 { get => field; }
+    public int P4 { get => field; set => field = value; }
+    public int P3 { get => field; set; }
+}
+```
+
+### Definite assignment in structs
+
+A semi-auto property will be treated as a regular auto property for the purposes of calculating default backing field initialization if its setter is automatically implemented, or if it does not have a setter ([LDM decision](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-02.md#property-assignment-in-structs)).
+
+Default initialize a struct when calling a manually implemented semi-auto property setter, and issue a warning when doing so, like a regular property setter ([LDM decision](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-05-02.md#definite-assignment-of-manually-implemented-setters)).
+
+### Nullability
+
+When `{ get; }` is written as `{ get => field; }`, or `{ get; set; }` is written as `{ get => field; set => field = value; }`, a similar warning should be produced when a non-nullable property is not initialized:
+
+```cs
+class C
+{
+    // ⚠️ CS8618: Non-nullable property 'P' must contain a
+    // non-null value when exiting constructor.
+    public string P { get => field; set => field = value; }
+}
+```
+
+No warning should be produced if the property is initialized to a non-null value via constructor assignment or property initializer:
+
+```cs
+class C
+{
+    public C() { P = ""; }
+
+    public string P { get => field; set => field = value; }
+}
+```
+
+```cs
+class C
+{
+    public string P { get => field; set => field = value; } = "";
+}
+```
+
+In the same vein as how `var` infers as nullable for reference types, the `field` type should be nullable for reference types. This makes sense of `field ??` as not being followed by dead code, and it avoids producing a misleading warning in the following example:
+
+```cs
+public string AmbientValue
+{
+    get => field ?? parent.AmbientValue;
+    set
+    {
+        if (value == parent.AmbientValue)
+            field = null; // No warning here. Resume following the parent's value.
+        else
+            field = value; // Stop following the parent's value
+    }
+}
+```
+
+Just like with nullability of `var`, it's expected that such patterns will exist with manually-declared backing fields.
+
+To land in this sweet spot implicitly, without having to write an attribute each time, nullability analysis will combine an inherent nullability of the field with the behavior of `[field: NotNull]`. This allows maybe-null assignments without warning, which is desirable as shown above, while simultaneously allowing a scenario like `=> field.Trim();` without requiring an intervention to silence a warning that `field` could be null. Making sure `field` has been assigned is already covered by the warning that ensures non-nullable properties are assigned by the end of each constructor.
```

Comment:
i think i'm good with this.  Though it doesn't mean you wouldn't get a warning here right:

```c#
public string AmbientValue
{
    get => field;  // Could return null now!
    set
    {
        if (value == parent.AmbientValue)
            field = null; // No warning here. Resume following the parent's value.
        else
            field = value; // Stop following the parent's value
    }
}
```

Might be worth calling this out, but with the caveat that we think the number of false negatives will be far fewer and less impactful than if we had too many false positives.

Reply:
You would not get that warning on `get => field;` because of `[field: NotNull]`. I'll test again to make sure I didn't make a mistake in testing that.

Reply:
right.  i'm saying: that's slightly problematic here, as the field *can* be null, and you could null ref.  But that we believe in aggregate the benefits here outweigh the problems.

Reply:
I think that scenario is well handled by the warning if you don't assign non-null properties in a constructor, which begins to apply here like it does with autoprops.

Reply:
Oops, I'm sorry, I didn't see at first that the `??` had been removed in the getter. I see your point: this is a bad setup and warnings would be nice, but we don't have a non-onerous way to provide them. I'll call that out.